File: client/computerpets_client/choice.py
"""A tap on the guest is a choice, not a sit. They pick. Then they do that sit.

Port of ``web/src/lib/pets/guest-choice.ts`` (and ``desktop/renderer/choice.js``).
"""

from __future__ import annotations

from typing import Iterable

GUEST_CHOICE = (
    "rest",
    "walk",
    "sit",
    "talk",
    "treat",
    "play",
    "special",
    "hide",
    "call",
    "pick",
)


def guest_tap() -> str:
    """A tap opens the choice. It does not talk, sleep, or walk them."""
    return "choice"


def pose_flip(walking: bool) -> dict[str, str]:
    """Walk if they are still. Sit if they are walking. One mark."""
    if walking:
        return {"id": "sit", "label": "Sit"}
    return {"id": "walk", "label": "Walk"}


def guest_marks(
    *,
    hidden: bool = False,
    leaving: bool = False,
    walking: bool = False,
    gifts: int = 0,
    treat_verb: str = "Treat",
    special_verb: str = "Special",
) -> list[dict[str, str]]:
    """The marks they may pick. Hidden keeps Call back. A gift on the wood keeps Pick."""
    busy = hidden or leaving
    marks: list[dict[str, str]] = []
    if not busy:
        marks.append({"id": "rest", "label": "Rest"})
        marks.append(pose_flip(walking))
    marks.append({"id": "talk", "label": "Talk"})
    if not busy:
        marks.append({"id": "treat", "label": treat_verb or "Treat"})
        marks.append({"id": "play", "label": "Play"})
    marks.append({"id": "special", "label": special_verb or "Special"})
    if hidden:
        marks.append({"id": "call", "label": "Call back"})
    elif not leaving:
        marks.append({"id": "hide", "label": "Hide"})
    if not hidden and gifts > 0:
        marks.append({"id": "pick", "label": "Pick"})
    return marks


def guest_pick(mark_id: str) -> str | None:
    """The sit they already know. A pick is not a new verb."""
    return mark_id if mark_id in GUEST_CHOICE else None


def guest_hit_pad(*, phone: bool = False, tablet: bool = False) -> int:
    """Extra wood around the guest on a finger sit. The pad is empty. It does not paint a plate."""
    if phone:
        return 12
    if tablet:
        return 16
    return 0


def walking_cmd(cmd: str | None) -> bool:
    return cmd in {"wander", "seek", "play", "enter"}


def mark_ids(marks: Iterable[dict[str, str]]) -> tuple[str, ...]:
    return tuple(m["id"] for m in marks)

File: desktop/renderer/choice.js
/** A tap on the guest is a choice, not a sit. They pick. Then they do that sit. */
(function (root) {
  const GUEST_CHOICE = [
    "rest",
    "walk",
    "sit",
    "talk",
    "treat",
    "play",
    "special",
    "hide",
    "call",
    "pick",
  ];

  function guestTap() {
    return "choice";
  }

  function poseFlip(walking) {
    return walking ? { id: "sit", label: "Sit" } : { id: "walk", label: "Walk" };
  }

  function guestMarks(sit) {
    sit = sit || {};
    const hidden = !!sit.hidden;
    const leaving = !!sit.leaving;
    const busy = hidden || leaving;
    const marks = [];
    if (!busy) {
      marks.push({ id: "rest", label: "Rest" });
      marks.push(poseFlip(!!sit.walking));
    }
    marks.push({ id: "talk", label: "Talk" });
    if (!busy) {
      marks.push({ id: "treat", label: sit.treatVerb || "Treat" });
      marks.push({ id: "play", label: "Play" });
    }
    marks.push({ id: "special", label: sit.specialVerb || "Special" });
    if (hidden) marks.push({ id: "call", label: "Call back" });
    else if (!leaving) marks.push({ id: "hide", label: "Hide" });
    if (!hidden && (sit.gifts || 0) > 0) marks.push({ id: "pick", label: "Pick" });
    return marks;
  }

  function guestPick(id) {
    return GUEST_CHOICE.indexOf(id) >= 0 ? id : null;
  }

  function guestHitPad(sit) {
    // Extra wood. The pad is empty. It does not paint a plate.
    if (sit && sit.phone) return 12;
    if (sit && sit.tablet) return 16;
    return 0;
  }

  const api = { GUEST_CHOICE, guestTap, poseFlip, guestMarks, guestPick, guestHitPad };
  if (typeof module !== "undefined" && module.exports) module.exports = api;
  root.PetChoice = api;
})(typeof window !== "undefined" ? window : globalThis);
